File: leetcode/239-sliding-window-maximum.py
import heapq


class Solution:
    def maxSlidingWindow(self, nums: [int], k: int) -> [int]:
        result = []
        q = [(-nums[i], i) for i in range(k)]
        heapq.heapify(q)
        result.append(-q[0][0])
        for i in range(k, len(nums)):
            heapq.heappush(q, (-nums[i], i))
            while q[0][1] <= i - k:
                heapq.heappop(q)
            result.append(-q[0][0])
        return result


nums = [1, 3, -1, -3, 5, 3, 6, 7]
k = 3
# print(Solution().maxSlidingWindow(nums, k))
# print(Solution().maxSlidingWindow([1], 1))
print(Solution().maxSlidingWindow([1, -1], 1))
# print(Solution().maxSlidingWindow([1, -1], 2))

File: leetcode/239-sliding-window-maximum.java
package leetcode;

import java.util.Arrays;
import java.util.PriorityQueue;

class Solution239 {
    public static void main(String[] args) {
        int[] data = new int[] { 1, 3, -1, -3, 5, 3, 6, 7 };
        System.out.println(Arrays.toString(new Solution239().maxSlidingWindow(data, 3)));
    }

    public int[] maxSlidingWindow(int[] nums, int k) {
        PriorityQueue<Integer> queue = new PriorityQueue<>(k);
        for (int i = 0; i < k; i++) {
            queue.add(-nums[i]);
        }
        int[] result = new int[nums.length - k + 1];
        for (int i = k; i < nums.length; i++) {
            System.out.println(queue);
            result[i - k] = -queue.peek();
            queue.remove(-nums[i - k]);
            queue.add(-nums[i]);
        }
        result[nums.length - k] = -queue.peek();
        return result;
    }
}
